src/colmap/estimators/cost_functions.h

```diff
@@ -84,16 +107,25 @@ class ReprojErrorCostFunction {
                             point3D_in_cam[0],
                             point3D_in_cam[1],
                             point3D_in_cam[2],
-                            &residuals[0],
-                            &residuals[1]);
-    residuals[0] -= T(observed_x_);
-    residuals[1] -= T(observed_y_);
+                            &residuals_ptr[0],
+                            &residuals_ptr[1]);
+    residuals_ptr[0] -= T(observed_x_);
+    residuals_ptr[1] -= T(observed_y_);
+    if (covariance_type_ == CovarianceType::GENERAL) {
```

Comment:
It feels like we can perform this check once in the Create(). Then, using templating/constexpr, we can return a cost function that has zero computational/storage overhead during the optimization.